blob/service.go

```diff
@@ -0,0 +1,308 @@
+			if err != nil {
+				resultErr[i] = fmt.Errorf("could not get the blob for the nID:%s. %s", nID.String(), err)
+				ch <- nil
+				return
+			}
+			ch <- blobs
+		}()
+	}
+
+	for range nIDs {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case b := <-ch:
+			if b == nil {
+				continue
+			}
+			blobs = append(blobs, b...)
+		}
+	}
+
+	if len(blobs) == 0 && len(resultErr) == 0 {
+		return nil, ErrBlobNotFound
+	}
+
+	err = errors.Join(resultErr...)
+	if err != nil {
+		err = fmt.Errorf("blob: %s", err)
+	}
+	return blobs, err
+}
+
+// Included verifies that the blob was included in a specific height.
+// To ensure that blob was included in a specific height, we need:
+// 1. verify the provided commitment by recomputing it;
+// 2. verify the provided Proof against subtree roots that were used in 1.;
+func (s *Service) Included(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	_ *Proof,
+	com Commitment,
+) (bool, error) {
+	// In the current implementation, LNs will have to download all shares to recompute the commitment.
+	// To achieve 1. we need to modify Proof structure and to store all subtree roots, that were
+	// involved in commitment creation and then call `merkle.HashFromByteSlices`(tendermint package).
+	// nmt.Proof is verifying share inclusion by recomputing row roots, so, theoretically, we can do
+	// the same but using subtree roots. For this case, we need an extra method in nmt.Proof
+	// that will perform all reconstructions,
+	// but we have to guarantee that all our stored subtree roots will be on the same height(e.g. one
+	// level above shares).
+	// TODO(@vgonkivs): rework the implementation to perform all verification without network requests.
+	_, _, err := s.getByCommitment(ctx, height, nID, com)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// getByCommitment retrieves the DAH row by row, fetching shares and constructing blobs in order to
+// compare Commitments. Retrieving is stopped once the requested blob/proof is found.
+func (s *Service) getByCommitment(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Blob, *Proof, error) {
+	log.Infow("requesting blob",
+		"height", height,
+		"nID", nID.String())
+
+	header, err := s.getByHeight(ctx, height)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		rawShares = make([]share.Share, 0)
+		proofs    = make(Proof, 0)
+		amount    int
+		blobShare *shares.Share
+	)
+
+	namespacedShares, err := s.sGetter.GetSharesByNamespace(ctx, header.DAH, nID)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, rowShares := range namespacedShares {
+		if len(rowShares.Shares) == 0 {
+			continue
+		}
+
+		rawShares = append(rawShares, rowShares.Shares...)
+		proofs = append(proofs, rowShares.Proof)
+
+		// reconstruct the `blobShare` from the first rawShare in range
+		// in order to get blob's length(first share will contain this info)
+		if blobShare == nil {
```

Comment:
I would leave here. It does not look complicated. But extracting will overcomplicate it:
we will have to pass and return a slice of rawShares(in case if there will be no padding, then it will be the same slice)+bShare+lenngth